src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -528,95 +549,350 @@ internal override void Complete()
+                                    {
+                                        ErrorRecord errorRecord;
+                                        if (isFirstError)
+                                        {
+                                            isFirstError = false;
+                                            // Produce a regular error record for the first line of the output
+                                            errorRecord = new ErrorRecord(new RemoteException(e.Data), "NativeCommandError", ErrorCategory.NotSpecified, e.Data);
+                                        }
+                                        else
+                                        {
+                                            // Wrap the rest of the output in ErrorRecords with the "NativeCommandErrorMessage" error ID
+                                            errorRecord = new ErrorRecord(new RemoteException(e.Data), "NativeCommandErrorMessage", ErrorCategory.NotSpecified, null);
+                                        }
+
+                                        _nativeProcessOutputQueue.Enqueue(new ProcessOutputObject(errorRecord, MinishellStream.Error));
+                                    }
                                 }
+                            });
+                            _nativeProcess.BeginErrorReadLine();
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Read the output from the native process and send it down the line.
+        /// </summary>
+        /// <returns>
+        /// True if there was any new input available, otherwise false.
+        /// </returns>
+        private bool ConsumeAvailableNativeProcessOutput()
+        {
+            bool isNewData = false;
+            if (_background == false)
+            {
+                if (_startInfo.RedirectStandardOutput || _startInfo.RedirectStandardError)
+                {
+                    ProcessOutputObject record;
+                    while (_nativeProcessOutputQueue.TryDequeue(out record))
+                    {
+                        ProcessOutputRecord(record);
+                        isNewData = true;
+                    }
+                }
+            }
+            return isNewData;
+        }
+
+        internal override void Complete()
+        {
+            try
+            {
+                if (_background == false)   
+                {
+                    //Wait for input writer to finish.
+                    _inputWriter.Done();
+
+                    //Wait for the process to exit and consume available output
+                    while (!_nativeProcess.HasExited)
+                    {
+                        // TODO: Currently the main pipeline Thread is just spinning.
+                        // It would be much better to park it until new input is available.
+                        ConsumeAvailableNativeProcessOutput();
```

Comment:
`ConcurrentQueue` has the same blocking interface: `Dequeue` (instead of `TryDequeue`), but it's not the point of this TODO note. We should not block, because process may just stop at any moment without producing output. The right way to do it would be to use wait/notify where notify will be called the code that adds to collection and by a callback about process finish.